index.d.ts

```diff
@@ -247,15 +247,33 @@ export type Observer<T> = {
   next?(value: T): void
 }
 
+/**
+ * Extend the state
+ *
+ * This is used by store enhancers and store creators to extend state.
+ * If there is no state extension, it just returns the state, as is, otherwise
+ * it returns the state joined with its extension.
+ */
+export type ExtendState<State, Extension> = [Extension] extends [never]
```

Comment:
Why the square brackets around `Extension` and `never`?

Reply:
this one is a real foot gun :)

It turns out that `type a = never extends never ? true : false` resolve a to `true`, but this:

```ts
type fancy<T> T extends never ? true : false
type a = fancy<never>
```

resolves to `false`. The reason is that T can be a union of stuff. So, to lock the type down to check for just `never`, we make it a tuple, and then the type inference works.

I'm working on the `applyMiddleware` failure now

Reply:
Well, that's quite quirky!

Reply:
> Well, that's quite quirky!

I'll say. Took me ages to find the right page to google. I think I'll find it again and add a link in the source

Reply:
That looks kinda like the "distributive inference" change or whatever it was that got added to RSK recently:

https://github.com/reduxjs/redux-starter-kit/pull/138

Reply:
precisely! https://github.com/reduxjs/redux-starter-kit/pull/138/files#diff-438bbd0288b07c3e7ae7f40f89f7f451R13